File: design-pattern/factory_method/factory_method.py
import abc

class Shape(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def draw(self):
        pass

class Circle(Shape):
    def draw(self):
        print('Circle class')

class Square(Shape):
    def draw(self):
        print('Square class')

class ShapeFactory:
    # use dictionary technical for java enum like
    # variables declare here is class's variable
    __shape = {}

    def __init__(self):
    ##    self.foo is class's property
        # TODO: check if instance initialized then return old instead create new one
        self.__shape['CIRCLE'] = Circle()
        self.__shape['SQUARE'] = Square()

    @staticmethod
    def getShape(self, shape):
        shape = shape.upper()

        if shape == 'CIRCLE':
            return Circle()
        elif shape == 'SQUARE':
            return Square()
        else:
            return None

s = ShapeFactory()
s.getShape('Circle').draw()
s.getShape('Square').draw()

s.setFoo()

a = ShapeFactory()
print(a.foo)

File: design-pattern/factory_method/factory_method.java
// interface
public interface Shape {
    void draw();
}

// 3 class implement Shape interface
public class Circle extends Shape {
    @override
    void draw() {
        System.out.print("Circle class");
    }
}

public class Square extends Shape {
    @override
    void draw() {
        System.out.print("Square class");
    }
}

public class Rectangle extends Shape {
    @override
    void draw() {
        System.out.print("Rectangle class");
    }
}

public enum Shapes {
    CIRCLE,
    SQUARE,
    RECTANGLE
}

// Factory method "sinh ra" class cụ thể dựa trên biến truyền vào
public class ShapeFactory {
    public Shape getShape(String Shape) {
        // TODO: check correctness operation upperCase
        swich (String.upperCase(shape)): {
            case Shapes.CIRCLE:
                return new Circle();
            case Shapes.SQUARE:
                return new Square();
            case Shapes.RECTANGLE:
                return new Rectangle();
            default:
                return null;
        }

    }
}

// Demo
public class Demo {
    public class void main(String[] args) {
        ShapeFactory shapeFactory = new ShapeFactory();

        Shape shape = shapeFactory.getShape('circle');
        circle.draw();

        shape = shapeFactory.getShape('square');
        shape.draw();

        shape = shapeFactory.getShape('rectangle');
        shape.draw();
    }
}
